https://m.blog.naver.com/PostView.nhn?blogId=moojigai07&logNo=120186757908&proxyReferer=https%3A%2F%2Fwww.google.com%2F

    1. 최소자승법으로 a,b구한 뒤 X_test의 Y값을 구한다 : y_pred
    -(a * X_test ) + b

    2. sklearn 회귀모델로 X_test의 Y값을 구한다 : y_pred_sk

    3. 실제값(y_test)과 y_pred, y_pred_sk 간의 차이를 구해서 얼마나 비슷하게 나오는지 비교

In [2]:
import pandas as pd
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from math import sqrt

In [4]:
product = pd.read_csv('product.csv', encoding='cp949')

In [6]:
len(product)

264

In [7]:
product_train = product[product.index < 200]
product_test = product[product.index >= 200]

In [9]:
X_train = product_train[['제품_친밀도']]
y_train = product_train.iloc[:,-1]

X_test = product_test[['제품_친밀도']]
y_test = product_test.iloc[:,-1]

In [10]:
linear = LinearRegression().fit(X_train, y_train)

In [12]:
predict_linear = linear.predict(X_test)

In [13]:
predict_linear

array([3.41419081, 3.09998587, 3.72839576, 3.41419081, 2.47157597,
       3.09998587, 2.47157597, 3.72839576, 3.09998587, 3.72839576,
       2.47157597, 3.41419081, 3.09998587, 3.72839576, 3.41419081,
       3.09998587, 3.09998587, 3.41419081, 3.41419081, 3.09998587,
       3.09998587, 3.41419081, 3.09998587, 2.47157597, 3.41419081,
       3.09998587, 3.09998587, 3.41419081, 3.09998587, 3.41419081,
       3.09998587, 3.41419081, 3.09998587, 3.41419081, 3.09998587,
       3.09998587, 3.41419081, 3.09998587, 3.09998587, 3.41419081,
       2.47157597, 3.09998587, 3.41419081, 3.41419081, 3.09998587,
       3.09998587, 3.41419081, 3.41419081, 3.41419081, 2.78578092,
       3.41419081, 3.09998587, 3.41419081, 3.09998587, 2.78578092,
       3.09998587, 3.09998587, 3.09998587, 2.47157597, 3.09998587,
       3.09998587, 3.41419081, 3.41419081, 2.78578092])

In [15]:
import numpy as np

In [23]:
product_train.head()

제품_친밀도  제품_적절성  제품_만족도
0       3       4       3
1       3       3       2
2       4       4       4
3       2       2       2
4       2       2       2

In [19]:
X_train.head()

제품_친밀도
0       3
1       3
2       4
3       2
4       2

In [28]:
y_train.head()

0    3
1    2
2    4
3    2
4    2
Name: 제품_만족도, dtype: int64

In [50]:
X_train.mean(), y_train.mean()

(제품_친밀도    2.825
 dtype: float64, 3.045)

In [30]:
X_train.sum(), y_train.sum()

(제품_친밀도    565
 dtype: int64, 609)

In [48]:
len(X_train), len(y_train)

(200, 200)

In [42]:
a = X_train.values
b = y_train.values

In [44]:
a * b

array([[ 9,  6, 12, ...,  9,  9, 12],
       [ 9,  6, 12, ...,  9,  9, 12],
       [12,  8, 16, ..., 12, 12, 16],
       ...,
       [ 9,  6, 12, ...,  9,  9, 12],
       [ 9,  6, 12, ...,  9,  9, 12],
       [12,  8, 16, ..., 12, 12, 16]], dtype=int64)

In [43]:
(a * b).sum()

344085

In [52]:
y_train[0]

3

In [58]:
X_train['제품_친밀도'][0]

3

In [80]:
y_train.mean(), X_train.mean()

(3.045, 제품_친밀도    2.825
 dtype: float64)

In [136]:
ja_list=[]
mo_list=[]

for i in range(len(X_train)):
    ja = (y_train[i]-3.045) * (X_train['제품_친밀도'][i]-2.825)
    ja_list.append(ja)
    mo = np.square((X_train['제품_친밀도'][i]-2.825))
    mo_list.append(mo)
    
a = sum(ja_list)/sum(mo_list)
a

0.3142049469964677

In [137]:
# for i in range(len(X_train)):
#     ja = (3.045-y_train[i]) * (X_train['제품_친밀도'][i]-2.825)
# ja

In [138]:
b = 3.045-(a*2.825)
b

2.1573710247349784

In [139]:
# train식의 : a, b
a,b 

(0.3142049469964677, 2.1573710247349784)

In [140]:
X_test.columns

Index(['제품_친밀도'], dtype='object')

In [141]:
y_pred_list=[]
for i in range(200, 264):
    y_pred = (a * X_test['제품_친밀도'][i]) + b
    y_pred_list.append(y_pred)
    
    
y_pred_list[0:10]

[3.4141908127208493,
 3.0999858657243817,
 3.728395759717317,
 3.4141908127208493,
 2.471575971731446,
 3.0999858657243817,
 2.471575971731446,
 3.728395759717317,
 3.0999858657243817,
 3.728395759717317]

In [143]:
y_test_list = list(y_test)

In [144]:
y_test[200]

4

In [145]:
y_pred_list[0:5]

[3.4141908127208493,
 3.0999858657243817,
 3.728395759717317,
 3.4141908127208493,
 2.471575971731446]

In [146]:
len(y_test_list), len(y_pred_list)

(64, 64)

In [147]:
y_test_list[0], y_pred_list[0]

(4, 3.4141908127208493)

In [148]:
for i in range(len(y_test_list)):
    result_cal = (np.square(y_test_list[i] - y_pred_list[i])).mean()
result_cal

0.04588981478105654

In [149]:
predict_linear[0]

3.414190812720848

In [150]:
for i in range(len(y_test_list)):
    result_model = (np.square(y_test_list[i] - predict_linear[i])).mean()
result_model

0.04588981478105616

### 회귀식으로 구한 y 값 - 실제 y값
 - 0.04588981478105654

### pred_linear 값 - 실제 y값
 - 0.04588981478105616